File: flexbox/src/androidTest/java/com/google/android/flexbox/FakeFlexContainer.java
/*
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.flexbox;

import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Fake implementation of {@link FlexContainer}.
 */
class FakeFlexContainer implements FlexContainer {

    private List<View> mViews = new ArrayList<>();

    private List<FlexLine> mFlexLines = new ArrayList<>();

    @FlexDirection
    private int mFlexDirection = FlexDirection.ROW;

    @FlexWrap
    private int mFlexWrap = FlexWrap.WRAP;

    @JustifyContent
    private int mJustifyContent = JustifyContent.FLEX_START;

    @AlignItems
    private int mAlignItems = AlignItems.STRETCH;

    @AlignContent
    private int mAlignContent = AlignContent.STRETCH;

    @Override
    public int getFlexItemCount() {
        return mViews.size();
    }

    @Override
    public View getFlexItemAt(int index) {
        return mViews.get(index);
    }

    @Override
    public View getReorderedFlexItemAt(int index) {
        return mViews.get(index);
    }

    @Override
    public void addView(View view) {
        mViews.add(view);
    }

    @Override
    public void addView(View view, int index) {
        mViews.add(index, view);
    }

    @Override
    public void removeAllViews() {
        mViews.clear();
    }

    @Override
    public void removeViewAt(int index) {
        mViews.remove(index);
    }

    @Override
    public int getFlexDirection() {
        return mFlexDirection;
    }

    @Override
    public void setFlexDirection(@FlexDirection int flexDirection) {
        mFlexDirection = flexDirection;
    }

    @Override
    public int getFlexWrap() {
        return mFlexWrap;
    }

    @Override
    public void setFlexWrap(@FlexWrap int flexWrap) {
        mFlexWrap = flexWrap;
    }

    @Override
    public int getJustifyContent() {
        return mJustifyContent;
    }

    @Override
    public void setJustifyContent(@JustifyContent int justifyContent) {
        mJustifyContent = justifyContent;
    }

    @Override
    public int getAlignContent() {
        return mAlignContent;
    }

    @Override
    public void setAlignContent(@AlignContent int alignContent) {
        mAlignContent = alignContent;
    }

    @Override
    public int getAlignItems() {
        return mAlignItems;
    }

    @Override
    public void setAlignItems(@AlignItems int alignItems) {
        mAlignItems = alignItems;
    }

    @Override
    public List<FlexLine> getFlexLines() {
        return mFlexLines;
    }

    @Override
    public int getDecorationLength(int childAbsoluteIndex, int childRelativeIndexInFlexLine,
            FlexItem flexItem) {
        return 0;
    }

    @Override
    public int getPaddingTop() {
        return 0;
    }

    @Override
    public int getPaddingLeft() {
        return 0;
    }

    @Override
    public int getPaddingRight() {
        return 0;
    }

    @Override
    public int getPaddingBottom() {
        return 0;
    }

    @Override
    public int getPaddingStart() {
        return 0;
    }

    @Override
    public int getPaddingEnd() {
        return 0;
    }

    @Override
    public int getChildWidthMeasureSpec(int widthSpec, int padding, int childDimension) {
        return ViewGroup.getChildMeasureSpec(widthSpec, padding, childDimension);
    }

    @Override
    public int getChildHeightMeasureSpec(int heightSpec, int padding, int childDimension) {
        return ViewGroup.getChildMeasureSpec(heightSpec, padding, childDimension);
    }

    @Override
    public int getLargestMainSize() {
        int largestSize = Integer.MIN_VALUE;
        for (FlexLine flexLine : mFlexLines) {
            largestSize = Math.max(largestSize, flexLine.mMainSize);
        }
        return largestSize;
    }

    @Override
    public int getSumOfCrossSize() {
        int sum = 0;
        for (int i = 0, size = mFlexLines.size(); i < size; i++) {
            FlexLine flexLine = mFlexLines.get(i);
            sum += flexLine.mCrossSize;
        }
        return sum;
    }

    @Override
    public void onNewFlexItemAdded(int childAbsoluteIndex, int childRelativeIndexInFlexLine,
            FlexLine flexLine) {
        // No op
    }

    @Override
    public void onNewFlexLineAdded(FlexLine flexLine) {
        // No op
    }

    @Override
    public void setFlexLines(List<FlexLine> flexLines) {
        mFlexLines = flexLines;
    }

    @Override
    public List<FlexLine> getFlexLinesInternal() {
        return mFlexLines;
    }
}

File: flexbox/src/androidTest/java/com/google/android/flexbox/FakeFlexContainer.kt
/*
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.flexbox

import android.view.View
import android.view.ViewGroup

/**
 * Fake implementation of [FlexContainer].
 */
internal class FakeFlexContainer : FlexContainer {

    private val views = mutableListOf<View>()

    private var flexLines = mutableListOf<FlexLine>()

    @FlexDirection
    private var flexDirection = FlexDirection.ROW

    @FlexWrap
    private var flexWrap = FlexWrap.WRAP

    @JustifyContent
    private var justifyContent = JustifyContent.FLEX_START

    @AlignItems
    private var alignItems = AlignItems.STRETCH

    @AlignContent
    private var alignContent = AlignContent.STRETCH

    override fun getFlexItemCount() = views.size

    override fun getFlexItemAt(index: Int) = views[index]

    override fun getReorderedFlexItemAt(index: Int) = views[index]

    override fun addView(view: View) {
        views.add(view)
    }

    override fun addView(view: View, index: Int) {
        views.add(index, view)
    }

    override fun removeAllViews() {
        views.clear()
    }

    override fun removeViewAt(index: Int) {
        views.removeAt(index)
    }

    override fun getFlexDirection() = flexDirection

    override fun setFlexDirection(@FlexDirection flexDirection: Int) {
        this.flexDirection = flexDirection
    }

    override fun getFlexWrap() = flexWrap

    override fun setFlexWrap(@FlexWrap flexWrap: Int) {
        this.flexWrap = flexWrap
    }

    override fun getJustifyContent() = justifyContent

    override fun setJustifyContent(@JustifyContent justifyContent: Int) {
        this.justifyContent = justifyContent
    }

    override fun getAlignContent() = alignContent

    override fun setAlignContent(@AlignContent alignContent: Int) {
        this.alignContent = alignContent
    }

    override fun getAlignItems() = alignItems

    override fun setAlignItems(@AlignItems alignItems: Int) {
        this.alignItems = alignItems
    }

    override fun getFlexLines() = flexLines

    override fun isMainAxisDirectionHorizontal(): Boolean {
        return flexDirection == FlexDirection.ROW || flexDirection == FlexDirection.ROW_REVERSE
    }

    override fun getDecorationLengthMainAxis(view: View, index: Int, indexInFlexLine: Int) = 0

    override fun getDecorationLengthCrossAxis(view: View) = 0

    override fun getPaddingTop() = 0

    override fun getPaddingLeft() = 0

    override fun getPaddingRight() = 0

    override fun getPaddingBottom() = 0

    override fun getPaddingStart() = 0

    override fun getPaddingEnd() = 0

    override fun getChildWidthMeasureSpec(widthSpec: Int, padding: Int, childDimension: Int): Int {
        return ViewGroup.getChildMeasureSpec(widthSpec, padding, childDimension)
    }

    override fun getChildHeightMeasureSpec(heightSpec: Int, padding: Int, childDimension: Int): Int {
        return ViewGroup.getChildMeasureSpec(heightSpec, padding, childDimension)
    }

    override fun getLargestMainSize() = flexLines.maxBy { it.mMainSize }?.mMainSize ?: Integer.MIN_VALUE

    override fun getSumOfCrossSize() = flexLines.sumBy { it.mCrossSize }

    override fun onNewFlexItemAdded(view: View, index: Int, indexInFlexLine: Int, flexLine: FlexLine) = Unit

    override fun onNewFlexLineAdded(flexLine: FlexLine) = Unit

    override fun setFlexLines(flexLines: List<FlexLine>) {
        this.flexLines = flexLines.toMutableList()
    }

    override fun getFlexLinesInternal() = flexLines

    override fun updateViewCache(position: Int, view: View) = Unit
}
